zerver/lib/narrow.py

```diff
@@ -755,11 +756,34 @@ def by_group_pm_with(
         return query.where(maybe_negate(cond))
 
     def by_search(self, query: Select, operand: str, maybe_negate: ConditionTransform) -> Select:
-        if settings.USING_PGROONGA:
+        if settings.USING_PG_SEARCH:
+            return self._by_search_pg_search(query, operand, maybe_negate)
+        elif settings.USING_PGROONGA:
             return self._by_search_pgroonga(query, operand, maybe_negate)
         else:
             return self._by_search_tsearch(query, operand, maybe_negate)
 
+    def _by_search_pg_search(
+        self, query: Select, operand: str, maybe_negate: ConditionTransform
+    ) -> Select:
+        # See search_operand_to_tantivy_query for details of what queries are supported.
+        text_search_query = search_operand_to_tantivy_query(operand)
+        safe_text_search_query = literal(text_search_query)
+
+        # TODO: finding the matches for higlighting is not implemented yet
+        query = query.add_columns(
+            literal_column("ARRAY[]::integer[]").label("content_matches"),
+            literal_column("ARRAY[]::integer[]").label("topic_matches"),
+        )
```

Comment:
`idxname.snippet(...)` can be used for higlighting (https://docs.paradedb.com/api-reference/full-text/highlighting#basic-usage), but it's a little bit tricky because it seems to escape HTML:

```
zulip=> SELECT * from fts_pg_search_idx.snippet('rendered_content:get', highlight_field => 'rendered_content', max_num_chars => 1300) where id=1009;
  id  |                                                                 snippet                                                                 | score_bm25
------+-----------------------------------------------------------------------------------------------------------------------------------------+------------
 1009 | &lt;p&gt;It&#x27;s nice and it &lt;em&gt;feels more modern, but I think&lt;/em&gt; this will take some time to <b>get</b> used to&lt;/p |  3.9742208
(1 row)

zulip=> select id, rendered_content from zerver_message where id=1009;
  id  |                                            rendered_content
------+---------------------------------------------------------------------------------------------------------
 1009 | <p>It's nice and it <em>feels more modern, but I think</em> this will take some time to get used to</p>
(1 row)
```

so that makes it a bit fiddly to be able to correctly map the highlighted positions in the snippet to the right positions in rendered_content. Probably we can just unescape the HTML again, but needs  testing to see if that behaves as well as i hope

Reply:
This seems like a question we should ask the ParadeDB folks about; I'd worry that offsets might be off if the original message contained stuff like `&nbsp;`. And it might be easy for them to just add a variant of this function that doesn't HTML-escape things.

Reply:
What we actually want are the `(position, length)` character offsets of each match (as a PostgreSQL multidimensional array `integer[][]`), so we can do the string processing ourselves. We don’t want the database to do the string processing at all.

Reply:
@andersk is your suggestion that we ask upstream to expose a function/option that'll return just the position offsets, instead of the snippet text with html tags inserted?

I think that'd indeed be best. Their current way of inserting html tags where the highlighting should be is consistent with how Postgres does (https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-HEADLINE) - which is why we do that painful processing in `ts_locs_array`. But getting the positions would obviously be superior. They rely on Tantivy (https://github.com/quickwit-oss/tantivy) to generate these snippets, so what they can do here might be constrained by the API tantivy offers, but we can ask.

Reply:
Yes, I think we should ask upstream for that; we got PGroonga to add something similar when we were integrating that, and I think we're not the only consumers of that API.

Reply:
We can expose this, it's been requested before. I've put up an issue for it: https://github.com/paradedb/paradedb/issues/1778

If one of you is feeling adventurous, we'd love a contribution! We may not be able to get to it super quickly otherwise.